content/en/docs/quarks-operator/concepts/ha.md

```diff
@@ -0,0 +1,182 @@
+
+
+# HA in Quarks
+
+In Quarks HA is implemented by using multiple availability zones, as described in the [qsts controller docs](../../../quarks-statefulset/development#az-support).
+
+In short, this uses labeled nodes and spawns `replica count` pods on each node.
+
+## Multi AZ
+
+In order for things to work correctly across versions and AZs, we need [ClusterIP `Services`](https://kubernetes.io/docs/tutorials/stateful-application/basic-stateful-set/#using-stable-network-identities) that select for Instance Group `Pods`.
+
+Services:
+```text
+nats-0
+  selects pod z0-v1-0
+  selects pod z0-v2-0
+nats-1
+  selects pod z1-v1-0
+  selects pod z1-v2-0
+nats-2
+  selects pod z0-v1-1
+  selects pod z0-v2-1
+nats-3
+  selects pod z1-v1-1
+  selects pod z1-v2-1
+nats-4
+  selects pod z0-v1-2
+  selects pod z0-v2-2
+nats-5
+  selects pod z1-v1-2
+  selects pod z1-v2-2
+```
+(from [Services And DNS Addresses](rendering_templates.md#services-and-dns-addresses) )
+
+The services select pods via the pod-ordinal label.
+
+Service selector example for "nats-0" service:
+```yaml
+selector:
+  quarks.cloudfoundry.org/az-index: "0"
+  quarks.cloudfoundry.org/deployment-name: kubecf
+  quarks.cloudfoundry.org/instance-group-name: nats
+  quarks.cloudfoundry.org/pod-ordinal: "0"
+```
+
+Quarks also generates an additional `nats' service.
+
+
+# Ordinals
+
+Each pod in an instance group's statefulset has several "ordinals" set as labels.
+
+* pod-ordinal
+* startup-ordinal
+* spec-index
+* az-index
+
+## Pod Ordinal
+
+Quarks stores the pod ordinal, which is the [suffix from the pod name](https://kubernetes.io/docs/tutorials/stateful-application/basic-stateful-set/#using-stable-network-identities), in a label.
+This is done by the pod mutator in Quarks statefulset.
+
+## Startup Ordinal
+
+Quarks also stores a startup ordinal in another label. Its value is 0 for the first pod to start in a statefulset revision. It is also persisted, so it stays the same if pods within a revision restart.
+
+That label is then referenced by the POD_ORDINAL env var in instance group pods. (There is special treatment for errands.)
+
+The POD_ORDINAL is used by the template-render sub command in an init container, to render the final version of the job configs from their eruby templates.
+
+## Spec Index
+
+Appears to be unused and is always equal to the pod ordinal.
+
+## AZ Index
+
+Used to calculate names and indexes for [multi AZ](from_bosh_to_kube.md#dealing-with-azs).
+
+It is calculated using the following formula:
+
+```text
+(AZ_INDEX - 1) * REPLICAS + POD_ORDINAL
+```
+
+# BOSH Job Spec
+
+Several BOSH mechanisms rely on the ordinals, which are derived from the statefulset state.
+
+I.e. the [job spec](https://bosh.io/docs/jobs/) is calculated with those numbers.
+
+Currently:
+
+* spec.bootstrap uses the startup-ordinal
+* spec.index uses the startup-ordinal
+* spec.address uses the startup-ordinal
```

Comment:
It's more correct, because it worked with bootstrap. It was believed to not matter much, but now we see the problems listed in the problems section.

* service points to wrong pod (possibly from another version)
* spec.address no longer matches hostname and cannot be used to bind to an interface